sys/vm_map.c

```diff
@@ -91,22 +94,31 @@ static bool vm_map_insert_entry(vm_map_t *vm_map, vm_map_entry_t *entry) {
 
 vm_map_entry_t *vm_map_find_entry(vm_map_t *vm_map, vm_addr_t vaddr) {
   vm_map_entry_t *etr_it;
-  TAILQ_FOREACH (etr_it, &vm_map->list, map_list)
-    if (etr_it->start <= vaddr && vaddr < etr_it->end)
+  rw_enter(&vm_map->rwlock, RW_READER);
+  TAILQ_FOREACH (etr_it, &vm_map->list, map_list) {
+    if (etr_it->start <= vaddr && vaddr < etr_it->end) {
+      rw_leave(&vm_map->rwlock);
       return etr_it;
+    }
+  }
+  rw_leave(&vm_map->rwlock);
   return NULL;
 }
 
 static void vm_map_remove_entry(vm_map_t *vm_map, vm_map_entry_t *entry) {
+  rw_assert(&vm_map->rwlock, RW_WLOCKED);
   vm_map->nentries--;
   vm_object_free(entry->object);
   TAILQ_REMOVE(&vm_map->list, entry, map_list);
   kfree(mpool, entry);
 }
 
 void vm_map_delete(vm_map_t *map) {
+  rw_enter(&map->rwlock, RW_WRITER);
```

Comment:
Aww... It'd be so nice to have magic macro to enable us to write:

```c
with_rw_lock(&map->rw_lock, RW_WRITER) {
 ...
}
```

... or alternatively use cleanup attribute (provided `rw_enter` returns pointer to locked object) like so:

```c
#define rw_auto_leave \
  rw_lock_t *__rw_ ## __LINE__ ## __attribute__((cleanup(rw_leave))) =
...
rw_auto_leave rw_enter(&map->rw_lock, RW_WRITER);
```